Entity/CustomItem.php

```diff
@@ -611,4 +620,39 @@ public function getRelationsByType(string $entityType): Collection
                 return new ArrayCollection([]);
         }
     }
+
+    public function getUniqueHash(): ?string
+    {
+        return $this->uniqueHash;
+    }
+
+    public function setUniqueHash(?string $uniqueHash): void
+    {
+        $this->uniqueHash = $uniqueHash;
+    }
+
+    public function createUniqueHash(): ?string
+    {
+        $rowData                = [];
+        $uniqueHash             = [];
+        $uniqueIdentifierFields = $this->customObject->getFieldsIsUniqueIdentifier();
+
+        if (is_null($uniqueIdentifierFields)) {
+            return null;
+        }
+
+        foreach ($this->getCustomObject()->getCustomFields()->getValues() as $customField) {
+            try {
+                $rowData[$customField->getAlias()] = $this->findCustomFieldValueForFieldAlias($customField->getAlias())->getValue();
+            } catch (NotFoundException $exception) {
```

Comment:
Is there need for this catch? I think that if we cannot find a value for a unique identifier field than it should throw an exception. But maybe I'm missing something?

Reply:
For this you'll have to look into the definition of `$this->findCustomFieldValueForFieldAlias`. It throws a `NotFoundException` when the value for a certain field is empty. 
So, when the loop iterates over a certain field the user has left empty, an exception is thrown and we don't want that in this case. Hence, this catch block is added to ignore that exception and move ahead

Reply:
Wouldn't we want to have the value set to an empty value in that case?

Reply:
We are only searching for a value of the unique identifier field, though, right? Not all of them. And each unique identifier field should be also required. So validation should catch that there is no value before we'll get to this code.

Or is it throwing this exception normally? I still think it shouldn't happen and if it would then it shouldn't be silently overlooked but it should throw an error.

Reply:
It really wouldn't matter. The $rowData variable is just storing the key value mappings of the field and the contained value. Something like {name => BMW, color => Black, id => 1234} Later in the next loop, we just filter out the unique identifier values (which can't be empty) from this variable and use them to form the unique hash.

If suppose, a user chooses to leave the color field blank, I'd get an exception here, which I don't want. I just want the iteration to go ahead and store the id. Because in the end I'd need only the unique identifier fields which definitely won't be empty

The reason I didn't do it in a single loop is because the only way I could think of required a nested loop. I found it to be a messy way to do it. So I went with 2 separate loops for storing key value pairs and fishing out unique values. Let me know if you think there's a better way to do this.

Reply:
How about to iterate directly on the unique fields? So instead of
```php
foreach ($this->getCustomObject()->getCustomFields()->getValues() as $customField) {
```
We'd do
```
foreach ($uniqueIdentifierFields->getValues() as $customField) {
```
And then you should be able to build the `$uniqueHash` array in this loop and won't need the other. Right?

I'd still suggest to remove the try/catch that may be silently hiding a bug. I suggest to move building of the hash after the validation of the entity:

https://github.com/acquia/mc-cs-plugin-custom-objects/pull/282/files#diff-08f0184015f98effc30ed5e9bfd5a2045cab1a49769bed0a35904358c4f782d1L107

So that we always get the valid data. All unique fields must be filled in as unique fields are automatically set as required.

I worry that if there is a bug somewhere and the exception is thrown then the custom item will update another existing custom item instead of creating a new one. This can lead to data loss which cannot be reverted. So I'd rather get a bug ticket that the exception is thrown and investigate why it is happening rather than losing silently data and get a bug report about that.

Reply:
I agree with @escopecz . 

Since the values should not be empty we should validate/transform the data instead of hiding potential bugs.

Reply:
Thanks a lot, those are really good insights and suggestions. I have refactored the function, could you take a look again?